language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/UsedTypeParams.scala

```diff
@@ -26,10 +31,191 @@ object UsedTypeParams {
       .foldMap(_.bifoldMap(f)(_.bifoldMap(_ foldMap (_.bifoldMap(notAGT)(f)))(f)))
   }
 
-  private def collectTypeParams(field: iface.Type): Set[String] = field match {
+  private[this] def collectTypeParams[S >: Ref.Name](field: iface.Type): Set[S] = field match {
     case iface.TypeVar(x) => Set(x)
     case iface.TypePrim(_, xs) => xs.toSet.flatMap(collectTypeParams)
     case iface.TypeCon(_, xs) => xs.toSet.flatMap(collectTypeParams)
     case iface.TypeNumeric(_) => Set.empty
   }
+
+  final case class LookupType[+RF, +VF](
+      run: iface.TypeConName => Option[(iface.DefDataType[RF, VF], LookupType[RF, VF])]
+  )
+
+  import VarianceConstraint.BaseResolution
+  import ScopedDataType.{Name => I}
+  import Variance._
+  private[this] type TVar = Ref.Name
+
+  final class ResolvedVariance private (private val prior: Map[I, ImmArraySeq[Variance]]) {
+    def allCovariantVars(
+        dt: I,
+        ei: iface.EnvironmentInterface,
+    ): (ResolvedVariance, ImmArraySeq[Variance]) = {
+      val resolved = covariantVars(dt, (i: I) => ei.typeDecls get i map (_.`type`))
+      (
+        resolved,
+        resolved.prior.getOrElse(
+          dt,
+          ei.typeDecls.getOrElse(dt, sys.error(s"$dt not found")).`type`.typeVars map (_ =>
+            Covariant
+          ),
+        ),
+      )
+    }
+
+    /** Variance of `sdt.typeVars` in order. */
+    private[this] def covariantVars[RF <: iface.Type, VF <: iface.Type](
+        dt: I,
+        lookupType: I => Option[iface.DefDataType[RF, VF]],
+    ): ResolvedVariance = {
+      import iface._
+
+      def lookupOrFail(i: I) = lookupType(i) getOrElse sys.error(s"$i not found")
```

Comment:
I think this error is far more likely to imply an invalid `EnvironmentInterface` (i.e. a bad LF) because every `dt` should have come from the `EnvironmentInterface`, and every reference within a data type therein should be resolvable.